masonry/src/widget/board.rs

```diff
@@ -0,0 +1,464 @@
+// Copyright 2018 the Xilem Authors and the Druid Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! A widget that arranges its children in fixed positions.
+
+use std::ops::{Deref as _, DerefMut as _};
+
+use accesskit::Role;
+use smallvec::SmallVec;
+use tracing::{trace_span, Span};
+use vello::Scene;
+
+use crate::widget::WidgetMut;
+use crate::{
+    AccessCtx, AccessEvent, BoxConstraints, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx,
+    Point, PointerEvent, Rect, Size, StatusChange, TextEvent, Widget, WidgetId, WidgetPod,
+};
+
+/// A container with absolute positioning layout.
+pub struct Board {
+    children: Vec<WidgetPod<Box<dyn SvgElement>>>,
+}
+
+/// Parameters for an item in a [`Board`] container.
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub struct BoardParams {
+    pub origin: Point,
+    pub size: Size,
+}
+
+/// Wrapper of a regular widget for use in [`Board`]
+pub struct PositionedElement<W> {
+    inner: W,
+    params: BoardParams,
+}
+
+/// A trait representing a widget which knows its origin and size.
+pub trait SvgElement: Widget {
+    // Origin of the widget relative to its parent
+    fn origin(&self) -> Point;
+    // Size of the widget
+    fn size(&self) -> Size;
+
+    // Sets the origin of the widget relative to its parent
+    fn set_origin(&mut self, origin: Point);
+    // Sets the size of the widget
+    fn set_size(&mut self, size: Size);
+}
```

Comment:
This trait strikes me as a symptom of trying way too hard to be clever in your design. Especially given that some implementations of the trait panic in some methods.

I think you could go pretty far by going a simpler route:

- No SvgElement trait.
- Both widgets and shapes are stored inside PositionedElement.
- PositionedElement doesn't implement Widget.
- The default way to add a shape is to add a PositionedElement with a KurboShape widget, an origin of `shape.bounding_box().origin()` and a size of  `shape.bounding_box().size()`.

Reply:
> This trait strikes me as a symptom of trying way too hard to be clever in your design.

That's mostly on me, the reason why I was pushing towards this, is to have strongly typed "border", to enforce only the "specialized" class of "svg" widgets within this kind of context (i.e. have a tree-hierarchy in near/mid-future).

I'd agree here, when this would only be one layer (similar as `Flex` etc.), but my plan with all of this is to have a completely separate (specialized) widget tree for the svg implementation.

I'll take a closer look in the following week and may use this PR as basis for a (xilem_)svg implementation.

> PositionedElement doesn't implement Widget.

The way I see `PositionedElement` is [`<foreignObject>`](https://developer.mozilla.org/en-US/docs/Web/API/SVGForeignObjectElement), and with all the previously said, I think it makes sense to have this implement widget as well.